Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -0,0 +1,750 @@
+    let _code: _Code
+
+    public static let alreadyFinished: Self = .init(_code: .alreadyFinished)
+}
+
+/// A ``NIOAsyncWriter`` is a type used to bridge elements from the Swift Concurrency domain into
+/// a synchronous world. The `Task`s that are yielding to the ``NIOAsyncWriter`` are the producers.
+/// Whereas the ``NIOAsyncWriterDelegate`` is the consumer.
+///
+/// Additionally, the ``NIOAsyncWriter`` allows the consumer to toggle the writability by calling ``NIOAsyncWriter/Sink/toggleWritability()``.
+/// This allows the implementation of flow control on the consumer side. Any call to ``NIOAsyncWriter/yield(contentsOf:)`` or ``NIOAsyncWriter/yield(_:)``
+/// will suspend if the ``NIOAsyncWriter`` is not writable and only be resumed after the ``NIOAsyncWriter`` becomes writable again.
+///
+/// - Note: It is recommended to never directly expose this type from APIs, but rather wrap it. This is due to the fact that
+/// this type has three generic parameters where at least two should be known statically and it is really awkward to spell out this type.
+/// Moreover, having a wrapping type allows to optimize this to specialized calls if all generic types are known.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public struct NIOAsyncWriter<
+    Element,
+    Failure,
+    Delegate: NIOAsyncWriterDelegate
+>: Sendable where Delegate.Element == Element, Delegate.Failure == Failure {
+    /// Simple struct for the return type of ``NIOAsyncWriter/makeWriter(elementType:failureType:isWritable:delegate:)``.
+    ///
+    /// This struct contains two properties:
+    /// 1. The ``sink`` which should be retained by the consumer and is used to toggle the writability.
+    /// 2. The ``writer`` which is the actual ``NIOAsyncWriter`` and should be passed to the producer.
+    public struct NewWriter {
+        public let sink: Sink
+        public let writer: NIOAsyncWriter
+
+        @usableFromInline
+        /* fileprivate */ internal init(
+            sink: Sink,
+            writer: NIOAsyncWriter
+        ) {
+            self.sink = sink
+            self.writer = writer
+        }
+    }
+
+    /// This class is needed to hook the deinit to observe once all references to the ``NIOAsyncWriter`` are dropped.
+    ///
+    /// If we get move-only types we should be able to drop this class and use the `deinit` of the ``NIOAsyncWriter`` struct itself.
+    ///
+    /// - Important: This is safe to be unchecked ``Sendable`` since the `storage` is ``Sendable`` and `immutable`.
+    @usableFromInline
+    /* fileprivate */ internal final class InternalClass: @unchecked Sendable {
+        @usableFromInline
+        internal let storage: Storage
+
+        @inlinable
+        init(storage: Storage) {
+            self.storage = storage
+        }
+
+        @inlinable
+        deinit {
+            storage.writerDeinitialized()
+        }
+    }
+
+    @usableFromInline
+    /* private */ internal let internalClass: InternalClass
```

Comment:
nit: we generally prefix with an underscore for private-but-usable-from-inline properties.